src/main/java/io/aiven/kafka/auth/TimedLockAclAuthorizer.java

```diff
@@ -0,0 +1,293 @@
+/*
+ * Copyright 2021 Aiven Oy https://aiven.io
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.aiven.kafka.auth;
+
+import java.io.File;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+
+import org.apache.kafka.common.security.auth.KafkaPrincipal;
+
+import io.aiven.kafka.auth.audit.AuditorAPI;
+import io.aiven.kafka.auth.json.AivenAcl;
+import io.aiven.kafka.auth.json.reader.AclJsonReader;
+import io.aiven.kafka.auth.json.reader.JsonReader;
+import io.aiven.kafka.auth.json.reader.JsonReaderException;
+
+import kafka.network.RequestChannel.Session;
+import kafka.security.auth.Acl;
+import kafka.security.auth.Authorizer;
+import kafka.security.auth.Operation;
+import kafka.security.auth.Resource;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+public class TimedLockAclAuthorizer implements Authorizer  {
+    private static final Logger LOGGER = LoggerFactory.getLogger(AivenAclAuthorizer.class);
+    private File configFile;
+    private ReentrantReadWriteLock lock = new ReentrantReadWriteLock(true);
+    private long lastUpdateCheckTimestamp = 0;
+    private long lastModifiedTimestamp = 0;
+    private List<AivenAcl> aclEntries;
+    private Map<String, Boolean> verdictCache;
+    private JsonReader<AivenAcl> jsonReader;
+    private AuditorAPI auditor;
+    private boolean logDenials;
+
+    public TimedLockAclAuthorizer() {
+    }
+
+    /**
+     * Reload ACLs from disk under a write lock.
+     */
+    public boolean reloadAcls() {
+        lock.writeLock().lock();
+        final long previousUpdate = lastModifiedTimestamp;
+        try {
+            checkAndUpdateConfig();
+            lastUpdateCheckTimestamp = System.nanoTime() / 1000000;
```

Comment:
Why do we don't use a currentTimeMillis?

Reply:
nanoTime is monotonic, never goes back

Reply:
Yes, however, if I remember correctly, it is not thread save, and in a line 150 it is requested out of lock....it looks weird